File: src/codec/encoder.cpp
#include "encoder.hpp"


namespace shar::codec {

Encoder::Encoder(Context context, Size frame_size, usize fps)
  : Context(context)
  , m_codec(std::move(context), frame_size, fps) {
}

void Encoder::run(Receiver<ffmpeg::Frame> input, Sender<ffmpeg::Unit> output) {
  Metric bytes_in{ m_metrics, "Encoder in", Metrics::Format::Bytes };
  Metric bytes_out{ m_metrics, "Encoder out", Metrics::Format::Bytes };

  while (auto frame = input.receive()) {
    if (m_running.expired() || !output.connected()) {
      break;
    }

    bytes_in += frame->total_size();
    auto units = m_codec.encode(std::move(*frame));
    for (auto& unit: units) {
      bytes_out += unit.size();
      output.send(std::move(unit));
    }

  }

  shutdown();
}

void Encoder::shutdown() {
  m_running.cancel();
}

}

File: src/codec/encoder.rs
use anyhow::Result;

use super::Unit;

pub trait Encoder: Send + 'static {
    type Frame: Send + 'static;
    type Unit: Unit;

    fn encode(&mut self, frame: Self::Frame, units: &mut Vec<Self::Unit>) -> Result<()>;
}

impl<E, F, U> Encoder for Box<E>
where
    E: Encoder<Frame = F, Unit = U> + ?Sized,
    F: Send + 'static,
    U: Unit,
{
    type Frame = F;
    type Unit = U;

    fn encode(&mut self, frame: Self::Frame, units: &mut Vec<Self::Unit>) -> Result<()> {
        <E as Encoder>::encode(self, frame, units)
    }
}
